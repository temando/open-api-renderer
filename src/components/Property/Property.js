import React, { PureComponent } from 'react'
import classNames from 'classnames'
import PropTypes from 'prop-types'
import Description from '../Description/Description'
import Indicator from '../Indicator/Indicator'
import PropertyConstraints from './PropertyConstraints/PropertyConstraints'
import { styles } from './Property.styles'

@styles
export default class Property extends PureComponent {
  constructor (props) {
    super(props)

    this.handleClick = this.handleClick.bind(this)
    this.displayAllEnums = this.displayAllEnums.bind(this)

    this.state = {
      enumValues: (props.enumValues ? props.enumValues : null)
    }
  }

  componentWillMount () {
    const enumValues = this.state.enumValues
    if (enumValues && enumValues.length > 20) {
      let updatedValues = enumValues.slice()
      updatedValues.length = 20
      this.setState({ enumValues: updatedValues })
    }
  }

  handleClick () {
    if (this.props.onClick) {
      this.props.onClick(this.props.name)
    }
  }

  render () {
    const {
      type, title, description, constraints, isRequired,
      defaultValue, onClick, isOpen, isLast,
      isAdditional, classes
    } = this.props

    let {name, enumValues} = this.props

    // If enumValues only has one single value, append the single value to name,
    // and do not display enum values
    if (enumValues && enumValues.length === 1) {
      name = `${name} = "${enumValues[0]}"`
      enumValues = []
    }

    const isClickable = onClick !== undefined

    let subtype
    if (type.includes('array')) {
      subtype = this.props.subtype
    }

    let indicatorDirection
    if (isOpen) {
      indicatorDirection = 'up'
    } else {
      indicatorDirection = 'down'
    }

    return (
      <tr
        className={classNames(classes.property, {
          [classes.last]: isLast
        })}
        onClick={this.handleClick}
      >
        <td className={classNames(classes.name, {
          [classes.isClickable]: isClickable,
          [classes.isAdditional]: isAdditional
        })}>
          <span>{name}</span>
          {isClickable && <Indicator className={classes.indicator} direction={indicatorDirection} />}
        </td>
        <td className={classes.info}>
          {title && <span className={classes.title}>{title}</span>}
          <span className={classes.type}>
            {!subtype ? type.join(', ') : <span className={classes.subType}>{subtype}[]</span>}
            {!subtype && constraints && constraints.format &&
            <span className={classes.format}>&lt;{constraints.format}&gt;</span>}
          </span>
          <PropertyConstraints constraints={constraints} type={type} isRequired={isRequired} />
          {((enumValues && enumValues.length) || defaultValue || description) &&
            <div className={classes.additionalInfo}>
              {enumValues && this.renderEnumValues(enumValues)}
              {defaultValue !== undefined && this.renderDefaultValue(defaultValue)}
              {description && <div><Description isInline description={description} /></div>}
            </div>
          }
        </td>
      </tr>
    )
  }

  /**
   * Render enum values
   *
   * @param {Array} values
   */
  renderEnumValues (values) {
    const { classes } = this.props
    const valuesToDisplay = this.state.enumValues
    let isEnumReduced = false
    if (values.length !== valuesToDisplay.length) {
      isEnumReduced = true
    }

    return (
      <div>
        <span>Valid values:</span>
        {valuesToDisplay.map(value => {
          return (
            <span key={value} className={classes.enum}>{value}</span>
          )
        })}
        {isEnumReduced &&
          <a onClick={this.displayAllEnums}>...</a>
        }
      </div>
    )
  }

  displayAllEnums () {
    this.setState({ enumValues: this.props.enumValues })
  }

  renderDefaultValue (value) {
    let displayValue

    if (typeof value === 'number' || typeof value === 'string') {
      displayValue = value
    } else if (typeof value === 'boolean') {
      displayValue = value.toString()
    } else if (Array.isArray(value)) {
      displayValue = `[${value.join(', ')}]`
    }

    return (
      <div className='default'>
        Default: <span>{displayValue}</span>
      </div>
    )
  }
}

Property.propTypes = {
  name: PropTypes.string.isRequired,
  type: PropTypes.arrayOf(PropTypes.string).isRequired,
  subtype: PropTypes.string,
  title: PropTypes.string,
  description: PropTypes.string,
  constraints: PropTypes.shape({
    format: PropTypes.string,
    exclusiveMinimum: PropTypes.number,
    exclusiveMaximum: PropTypes.number,
    maximum: PropTypes.number,
    maxItems: PropTypes.number,
    maxLength: PropTypes.number,
    maxProperties: PropTypes.number,
    minimum: PropTypes.number,
    minItems: PropTypes.number,
    minLength: PropTypes.number,
    minProperties: PropTypes.number,
    multipleOf: PropTypes.number,
    pattern: PropTypes.string,
    uniqueItems: PropTypes.bool
  }),
  enumValues: PropTypes.array,
  defaultValue: PropTypes.any,
  isRequired: PropTypes.bool,
  isOpen: PropTypes.bool,
  isLast: PropTypes.bool,
  isAdditional: PropTypes.bool,
  onClick: PropTypes.func,
  classes: PropTypes.object
}
