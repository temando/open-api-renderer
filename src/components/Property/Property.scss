@import '../../base.scss';

$tree-line-width: 1px;
$tree-border-style: $tree-line-width solid red;
$line-height: 40px;
$cell-padding: 10px;

.property-name,
.property-info {
  vertical-align: middle;
  padding-right: $cell-padding;
}

.property-name {
  width: 1%;
  white-space: nowrap;
  position: relative;
  span {
    display: inline-block;
    padding-right: $cell-padding;
    line-height: $line-height;
    vertical-align: middle;
  }
}

.property-info {
  width: 75%;
}

.property+.property {
  .property-info {
    border-top: 1px solid $border-color;
  }
}

.property-required {
  margin-left: 10px;
  color: red;
}

.property--isclickable {
  cursor: pointer;
  font-weight: bold;
}

/* Tree */

.property+.property, .body-schema-subset+.property {
  .property-name::after {
    content: '';
    border-left: $tree-border-style;
    position: absolute;
    bottom: 0;
    left: 0;
    height: 100%;
  }
}

.body-schema-subset > td {
  border-left: $tree-border-style;
}

.property.last+.body-schema-subset > td {
  border-left: none;
}

.property-name {
  &::before {
    content: '';
    display: inline-block;
    vertical-align: middle;
    border-top: $tree-border-style;
    width: 2 * $cell-padding;
  }
  span:first-child::before {
    content: '';
    display: inline-block;
    width: $cell-padding;
    height: $tree-line-width + 7;
    vertical-align: middle;
    border-left: $tree-border-style;
    position: relative;
    bottom: 2px;
  }
}

.property {
  &:first-child {
    .property-name::after {
      content: '';
      border-left: $tree-border-style;
      position: absolute;
      bottom: 0;
      left: 0;
      height: 52%;
    }
  }
  &:last-child, &.last {
    .property-name::after {
      content: '';
      border-left: $tree-border-style;
      position: absolute;
      top: 0;
      left: 0;
      height: 49%;
    }
  }
  &:only-of-type {
    .property-name::after {
      border: none;
    }
  }
}

.property {
  &:first-child.last {
    .property-name::after {
      border: none;
    }
  }
}

.enum,
.default {
  display: inline-block;
  padding: 0 5px;
  margin: 2px 3px;
  border: 1px solid grey;
}
