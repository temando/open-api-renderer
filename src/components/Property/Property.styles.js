import { createSheet } from '../../theme'
import c from 'color'

const lineHeight = '40px'
const cellPadding = '10px'

const nameAndInfo = {
  verticalAlign: 'middle',
  paddingRight: `${cellPadding}`
}

export const styles = createSheet(({ borders, backgrounds, text }) => ({
  'name': {
    ...nameAndInfo,
    width: '1%',
    whiteSpace: 'nowrap',
    position: 'relative',

    '&::before': {
      content: '""',
      display: 'inline-block',
      verticalAlign: 'middle',
      borderTop: `1px solid ${borders.default}`,
      width: '20px'
    },

    '& span': {
      display: 'inline-block',
      paddingRight: `${cellPadding}`,
      lineHeight,
      verticalAlign: 'middle',

      '&:first-child::before': {
        content: '""',
        display: 'inline-block',
        width: `${cellPadding}`,
        height: `8px`,
        verticalAlign: 'middle',
        borderLeft: `1px solid ${borders.default}`
      }
    }
  },

  'info': {
    ...nameAndInfo,
    width: '75%',
    paddingTop: '5px'
  },

  'additionalInfo': {
    fontSize: '0.9em',
    color: `${c(text.secondary)}`,
    padding: '5px 0',

    '& a': {
      cursor: 'pointer'
    }
  },

  'constraints': {
    marginLeft: '5px',
    fontSize: '.9rem',
    '& > span > span': {
      padding: '2px',
      backgroundColor: `${c(borders.secondary)}`,
      marginLeft: '5px'
    }
  },

  'property': {
    '&:first-child': {
      '& $name::after': {
        content: '""',
        borderLeft: `1px solid ${borders.default}`,
        position: 'absolute',
        bottom: '0',
        left: '0',
        height: '51%'
      }
    },
    '&:last-child, &$last': {
      '& $name::after': {
        content: '""',
        borderLeft: `1px solid ${borders.default}`,
        position: 'absolute',
        top: '0',
        left: '0',
        height: '50%'
      }
    },
    '&:only-of-type': {
      '& $name::after': {
        border: 'none'
      }
    },
    '&:first-child$last': {
      '& $name::after': {
        border: 'none'
      }
    },
    '& + &': {
      '& $info': {
        borderTop: `1px solid ${borders.default}`
      },
      '& $name::after': {
        content: '""',
        borderLeft: `1px solid ${borders.default}`,
        position: 'absolute',
        bottom: '0',
        left: '0',
        height: '100%'
      }
    }
  },

  'required': {
    marginLeft: '10px',
    color: 'red'
  },

  'isClickable': {
    cursor: 'pointer',
    fontWeight: 600
  },

  'isAdditional': {
    fontStyle: 'italic'
  },

  'last': {},

  'enum': {
    display: 'inline-block',
    padding: '0 5px',
    margin: '2px 3px',
    border: `1px solid ${borders.strong}`
  },
  'default': {
    extend: 'enum'
  },
  'indicator': {},
  'format': {},
  'subType': {}
}))
