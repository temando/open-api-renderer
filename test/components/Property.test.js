import React from 'react'
import Property from './../../src/components/Property/Property'
import renderer from 'react-test-renderer'
import ReactShallowRenderer from 'react-test-renderer/shallow'

describe('<Property />', () => {
  it('can render a basic property', () => {
    const tree = renderer.create(
      <Property name={'type'} type={['string']} isRequired isLast />
    )

    expect(tree).toMatchSnapshot()
  })

  it('can render a property with enum', () => {
    const tree = renderer.create(
      <Property
        name={'packagingType'}
        type={['string']}
        enumValues={['box', 'carton']}
        isRequired />
    )

    expect(tree).toMatchSnapshot()
  })

  it('can render a property with a subtype', () => {
    const tree = renderer.create(
      <Property
        name={'data'}
        type={['array']}
        subtype={'object'}
        isRequired={false} />
    )

    expect(tree).toMatchSnapshot()
  })

  it('can render a property with description', () => {
    const shallow = new ReactShallowRenderer()
    const tree = shallow.render(
      <Property
        name={'type'}
        type={['string']}
        description={'This is _markdown_ text'}
        isRequired />
    )

    expect(tree).toMatchSnapshot()
  })

  it('can render a property with numerical constraints', () => {
    const shallow = new ReactShallowRenderer()
    const tree = shallow.render(
      <Property
        name={'type'}
        type={['number']}
        constraints={{multipleOf: 2}}
        isRequired />
    )

    expect(tree).toMatchSnapshot()
  })

  it('can render a property with multiple types', () => {
    const tree = renderer.create(
      <Property
        name={'value'}
        type={['string', 'number']}
        isRequired={false} />
    )

    expect(tree).toMatchSnapshot()
  })

  it('can render a property with a format', () => {
    const tree = renderer.create(
      <Property
        name={'value'}
        constraints={{format: 'email'}}
        type={['string']}
        isRequired={false} />
    )

    expect(tree).toMatchSnapshot()
  })
})
