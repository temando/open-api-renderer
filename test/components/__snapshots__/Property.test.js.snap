// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Property /> can render a basic property 1`] = `
<tr
  className="undefined"
  onClick={[Function]}
>
  <td
    className=""
  >
    <span>
      type
    </span>
  </td>
  <td
    className={undefined}
  >
    <span
      className={undefined}
    >
      string
    </span>
    <span
      className={undefined}
    >
      <span
        className={undefined}
      >
        required
      </span>
    </span>
  </td>
</tr>
`;

exports[`<Property /> can render a property with a format 1`] = `
<tr
  className=""
  onClick={[Function]}
>
  <td
    className=""
  >
    <span>
      value
    </span>
  </td>
  <td
    className={undefined}
  >
    <span
      className={undefined}
    >
      string
      <span
        className={undefined}
      >
        &lt;
        email
        &gt;
      </span>
    </span>
    <span
      className={undefined}
    />
  </td>
</tr>
`;

exports[`<Property /> can render a property with a subtype 1`] = `
<tr
  className=""
  onClick={[Function]}
>
  <td
    className=""
  >
    <span>
      data
    </span>
  </td>
  <td
    className={undefined}
  >
    <span
      className={undefined}
    >
      <span
        className={undefined}
      >
        object
        []
      </span>
    </span>
    <span
      className={undefined}
    />
  </td>
</tr>
`;

exports[`<Property /> can render a property with description 1`] = `
<tr
  className=""
  onClick={[Function]}
>
  <td
    className=""
  >
    <span>
      type
    </span>
  </td>
  <td
    className={undefined}
  >
    <span
      className={undefined}
    >
      string
    </span>
    <span
      className={undefined}
    >
      <span
        className={undefined}
      >
        required
      </span>
    </span>
    <div
      className={undefined}
    >
      <div>
        <div
          className="undefined"
          dangerouslySetInnerHTML={
            Object {
              "__html": "This is <em>markdown</em> text",
            }
          }
        />
      </div>
    </div>
  </td>
</tr>
`;

exports[`<Property /> can render a property with docs 1`] = `
<tr
  className=""
  onClick={[Function]}
>
  <td
    className=""
  >
    <span>
      value
    </span>
  </td>
  <td
    className={undefined}
  >
    <span
      className={undefined}
    >
      string
    </span>
    <span
      className={undefined}
    />
    <div
      className={undefined}
    >
      <div
        className=""
      >
        <div
          className=""
          dangerouslySetInnerHTML={
            Object {
              "__html": "<p>Hello</p>
",
            }
          }
        />
        <a
          href="http://example.org"
          rel="noopener noreferrer"
          target="_blank"
        >
          More information Â»
        </a>
      </div>
    </div>
  </td>
</tr>
`;

exports[`<Property /> can render a property with enum 1`] = `
<tr
  className=""
  onClick={[Function]}
>
  <td
    className=""
  >
    <span>
      packagingType
    </span>
  </td>
  <td
    className={undefined}
  >
    <span
      className={undefined}
    >
      string
    </span>
    <span
      className={undefined}
    >
      <span
        className={undefined}
      >
        required
      </span>
    </span>
    <div
      className={undefined}
    >
      <div>
        <span>
          Valid values:
        </span>
        <span
          className={undefined}
        >
          box
        </span>
        <span
          className={undefined}
        >
          carton
        </span>
      </div>
    </div>
  </td>
</tr>
`;

exports[`<Property /> can render a property with multiple types 1`] = `
<tr
  className=""
  onClick={[Function]}
>
  <td
    className=""
  >
    <span>
      value
    </span>
  </td>
  <td
    className={undefined}
  >
    <span
      className={undefined}
    >
      string, number
    </span>
    <span
      className={undefined}
    />
  </td>
</tr>
`;

exports[`<Property /> can render a property with numerical constraints 1`] = `
<tr
  className=""
  onClick={[Function]}
>
  <td
    className=""
  >
    <span>
      type
    </span>
  </td>
  <td
    className={undefined}
  >
    <span
      className={undefined}
    >
      number
    </span>
    <span
      className={undefined}
    >
      <span
        className={undefined}
      >
        required
      </span>
      <span>
        <span
          className={undefined}
        >
          multiple of 2
        </span>
      </span>
    </span>
  </td>
</tr>
`;
